### DISCUSSION TITLE
Tailwind v4

### DISCUSSION THREAD

**Discussion by kuchta** [saadeghi/daisyui#2894] [General]
Hi guys,

I'm very excited about new [Tailwind v4](https://tailwindcss.com/blog/tailwindcss-v4-alpha) developments and since I know there's no supports for plugins yet, do you think daisyui could be implemented with just the new "css-first" approach without using it as a plugin?

**Comment by saadeghi** [comment]
It's mentioned in the blog that plugins will be supported in the stable release (using the config file)

I'm also excited about getting rid of JS config file.
Hopefully importing plugins will also be possible using the CSS file. It will be great for daisyUI because daisyUI source files are CSS but I have to convert them all to CSS in JS during build time just so it can be imported in the JS file.

**Reply by kuchta** [reply]
Well, that's exactly my question, since the daisyUI is just CSS is should *already* be possible to @import it (with some modifications) even without plugin support, right? I think we should leave this discussion open so other people playing with it can report any success here. I'm probably going to play with it as well. Is there something that you know could cause problems?

**Reply by kuchta** [reply]
The thing is untill the stable version is released, we still have a change to somehow influence it's design if it doesn't fit this project. Then it will be too late...

**Reply by saadeghi** [reply]
Importing a CSS file is already possible (native CSS)  
But what we would need is the imported CSS to be processed by Tailwind and unused styles to be ignored. Otherwise all the unused styles will be in the production which is not efficient.

daisyUI is not the only plugin for Tailwind CSS. There are many plugins out there and it will be their challenge as well. I think if Tailwind team is going to move forward to use the CSS file without a JS config, they will consider this as well.  
If not, plugins would need to be imported in the JS file (the current method) and it will still work. It will still be possible to use the JS config file as mentioned in the blog post.

**Reply by kuchta** [reply]
With this alpha version, we should *already* be able to just import daisyUI from CSS file and it should be processed by Tailwind automaticaly (Tailwind has now it's own vite plugin) including ignoring unused styles and without that build step converting CSS to JS. No suport for plugins is needed for that. My few projects that use daisyUI don't use any other external Tailwind plugins so I'm not that much interested in plugin support and I believe I'm not alone. The point of this discussion is to be a place where people can share their experience with this new approch so it would be nice of you to re-open it, even if you are at the moment not interested in it.

**Reply by saadeghi** [reply]
"Closing" is not equal to "Locking" and it doesn't mean I'm not interested or people can't have a discussion here.  
Open discussions are the ones that are still waiting for an answer. This one asked for my opinion and I answered. Of course anyone still can have a conversation here, shar their opinions or solutions, etc. I didn't lock it.

**Reply by saadeghi** [reply]
As I said, I'm also excited about it and I'm also waiting to see how the stable API for Tailwind 4 would look like. And I'm ready to realize daisyUI 5 if it needs breaking changes to br aligned with Tailwind 4

**Reply by kuchta** [reply]
Well, the problem with closed discussions is that they are not shown at the main page, so people wouldn't notice until they intentionally search for them and I actually wanted an answer from somebody who has practical experience with it...

**Reply by saadeghi** [reply]
>I actually wanted an answer from somebody who has practical experience with it...

What's the question? 

>do you think daisyui could be implemented with just the new "css-first" approach without using it as a plugin?

I answered your question. Currently it's not implemented in the alpha version. There no API for it and there's no docs yet.

But this is the goal according to Adam's tweet:

![Screenshot_2024-03-10-15-28-06-312_com twitter android](https://github.com/saadeghi/daisyui/assets/7342023/04f347e0-a49b-4055-95c2-3dc624ccb608)

**Reply by kuchta** [reply]
What's the question? Have somebody practical experience using (porting) daisyUI to the Tailwind v4 alpha?

If you had at least taken the time to understand the the blog post, you would't close the discussion, because you would understand that if as you say daisyUI is just css, than you probably don't need the new plugin API at all! You can just import daisyUI css and port all theme() functions calls to css variables and I don't know what else... That's why I'm asking, because I don't know daisyUI internals and it's interface with Tailwind to answer that, but you evidently don't know what this new version brings and your answers reflect that. Sorry to say that, but it's getting a little bit unbearable...

**Reply by saadeghi** [reply]
>Have somebody practical experience using (porting) daisyUI to the Tailwind v4 alpha?

How can possibly anyone have practical experience of using a Tailwind CSS plugin when the creator of Tailwind CSS said it DOES NOT EXIST in the current alpha release?

>we should leave this discussion open

>the problem with closed discussion

>you would't close the discussion

This is why you're a angry.

I've never seen anyone being this obsessive about a GitHub discussion so I don't know what to say at this point...

**Reply by kuchta** [reply]
If you understand the design of the new version, you wouldn't again and again talk about daisyUI as a plugin of Tailwind anymore. That's where the misunderstandings occur. If you don't know what to add to the discussion, it would be probably better not say anything and let it open for others to say samething actually usefull. I know this is you project and your repo, but you are not the only one here...

**Comment by oneezy** [comment]
hey guys! @saadeghi @kuchta 

it is indeed possible to easily import existing css in Tailwind v4 and everything just work.

if you need to overwrite/extend or create new TW v4 utility classes then it's easy as:
```css
/* app.css */
@import 'tailwdind.css';

@theme {
  color-green-500: #bada55;
  color-primary: #bada55;
}
```
that's it! 
I've been hacking away at my own little v4 ui lib as a proof of concept and i haven't ran into any issue.
- https://github.com/oneezy/ui/blob/next/src/lib/ui.css `@import`'s
- https://github.com/oneezy/ui/blob/next/src/lib/css/colors/colors.css `@theme`

my ui lib is built for svelte v5/ sveltekit but feel free to install it:
```
pnpm i -D @oneezy/ui@next
```
I feel your pain @kuchta , I've been been urging other ui lib maintainers they can begin updating to v4 with no problems.

 so i would say jump on it @saadeghi 😎

**Reply by saadeghi** [reply]
Thanks for the suggestion, but unfortunately this is not different from simply importing the whole [full.css](https://unpkg.com/daisyui@4.11.1/dist/full.css) file. All the unused styles are included with this method which is not efficient for production.

**Reply by oneezy** [reply]
Whatever you wrap inside `@theme` will only be created if it's used so I'm not sure what you're referring to

**Reply by saadeghi** [reply]
@oneezy The components, all the daisyUI class names,… we can't put them inside `@theme`:   
https://play.tailwindcss.com/lv4EYb3reH?file=css   
```diff
- `@theme` blocks must only contain custom properties or `@keyframes`.
```

**Reply by oneezy** [reply]
Oh, you guys don't even include the semantic classes in by default? in other words, if I install DaisyUI and never use the `.btn` class, it will never be generated?

Here' an example of what I was referring to:
https://play.tailwindcss.com/Ak17xxMmTb?file=css

`<button class="btn">button</button>`

![image](https://github.com/saadeghi/daisyui/assets/7369575/62df6663-4bcf-4d95-89c8-fac4b0e7dff7)
```css
@import "tailwindcss";

@theme {
  --color-primary: var(--color-blue-500, blue);
  --color-primary-content: var(--color-white, white);
}

@layer components {
  .btn {
    @apply bg-primary text-primary-content;
  }
}
```

**Reply by saadeghi** [reply]
> Oh, you guys don't even include the semantic classes in by default? in other words, if I install DaisyUI and never use the `.btn` class, it will never be generated?

Only used class names goes to the production CSS file.  Otherwise all the unused class names with all possible variations will be a big file.
Just like other Tailwind CSS class names. For example if `m-10` is not used, it won't be in CSS file.  
And because we are using the [Tailwind CSS plugins API](https://tailwindcss.com/docs/plugins#adding-components), it works the same for daisyUI as a plugin.
Currently the plugins API is not implemented in Tailwind CSS 4-alpha. But it's just alpha. It's not ready for production. Even all the CSS variables you don't use are included in the production file as you can see in the `Generated CSS` section in Tailwind Play…

If you can't wait for the stable version of Tailwind CSS, a workaround would be using daisyUI separately, as a CSS file and remove the unused style using PurgeCSS

**Reply by oneezy** [reply]
> And because we are using the [Tailwind CSS plugins API](https://tailwindcss.com/docs/plugins#adding-components), it works the same for daisyUI as a plugin.

ahhh ok that makes sense on why you're holding off. i wasn't aware of the ability to tree shake out your own custom classes you create. good to know!

**Comment by saadeghi** [comment]
I started this discussion for daisyUI 5 alpha: 
https://github.com/saadeghi/daisyui/discussions/3246

Let me if you have any feedbacks, suggestions, questions
